lib/chef/provider/package/snap.rb

```diff
@@ -138,10 +138,48 @@ def call_snap_api(method, uri, post_data = nil?)
           if defined?(::UNIXSocket)
             UNIXSocket.open("/run/snapd.socket") do |socket|
               # Send request, read the response, split the response and parse the body
-              socket.print(request)
-              response = socket.read
-              headers, body = response.split("\r\n\r\n", 2)
-              JSON.parse(body)
+              socket.write(request)
+
+              # WARNING!!! HERE BE DRAGONS
+              #
+              # So snapd doesn't return an EOF at the end of its body, so
+              # doing a normal read will just hang forever.
+              #
+              # Well, sort of. If, after it writes everything, you then send
+              # yet-another newline, it'll then send its EOF and promptly
+              # disconnect closing the pipe and preventing reading. So, you
+              # have to read first, and therein lies the EOF problem.
+              #
+              # The only way I've found to reliably not hang is read a character
+              # at a time until we have JSON that parses.
+              #
+              # Reading a character at a time isn't efficient, and since we know
+              # that HTTP headers always have a blank line after them, we can
+              # read lines until we find a blank line and *then* read a
+              # character at a time. snap returns all the JSON on a single line,
+              # so once you pass headers you MUST read a character a time.
+              #
+              # - jaymzh
+              Chef::Log.trace("snap_package[#{new_resource.package_name}]: Reading headers")
+              loop do
+                response = socket.readline
+                break if response.strip.empty? # finished headers
+              end
+              Chef::Log.trace("snap_package[#{new_resource.package_name}]: Past headers, onto the body...")
+              result = nil
+              body = ''
+              socket.each_char do |c|
+                body << c
+                begin
+                  result = JSON.parse(body)
+                  # if we get here, we were able to parse the json so we
+                  # are done reading
+                  break
+                rescue JSON::ParserError
+                  next
+                end
+              end
+              result
```

Comment:
If `read_nonblock` and friends don't work or are undesirable here, since a valid json document will end with a `]` or a `}` (right?), this could read until one of those characters is found before attempting to parse json.

Reply:
Nice optimization. Added.
Also note that in my new description I found you *can* make `read_nonblock` work, but it's 10x slower than this. ANd that was before your optimization.